### ISSUE TITLE
Issues installing on Linux with nvidia

### ISSUE THREAD

**Issue by tiagoantao** [mikeseven/node-opencl#55]
When I installed node-opencl and run the examples, I got an immediate segmentation fault.
My configuration is as follows:
- Ubuntu 16.04
- Nvidia Tesla K40
- NVidia OpenCL bindings that are v 1.2 (not 2)

It so happens that the opencl headers that are made available with ubuntu (package opencl-headers) are 2.0

This means that, for example, commandqueue.cpp `#ifndef CL_VERSION_2_0` will add the wrong function on compilation.

My hack was to edit the headers and undefine CL_VERSION_2_0

I do not have the experience to make a recommendation here, maybe this is just a problem with the way things are set up on ubuntu, but I think one of these might work

* document this so that people are aware OR
* detect the openCL version of the device in a different, more dynamic way

**Comment by robertleeplummerjr** [comment]
Did you get any further here?

**Comment by mikeseven** [comment]
Sorry, I was caught up in too many urgent things. I may be able to spend some time next week with holidays.

Your solution is indeed a good one I used in the past in other projects. Being more dynamic is better if you only use binaries but I don't use that as I prefer the native part to be compiled for your system. This ensures you are installing with a proper runtime, which is sometimes tricky to setup on some distributions.

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
It looks like this library will be used by http://gpu.rocks, fyi.  You think there will be sufficient support?

**Comment by mikeseven** [comment]
Cool, I was not aware of this library. I'll look at it.
At first glance, it seems to be a wrapper that would fit on top of node-opencl ;-)


--mike

________________________________

**Comment by robertleeplummerjr** [comment]
exactly.  the architectural layer would give us a means of finally having a full on GPU + js api.  Design in simple js, execute... everywhere.

**Comment by robertleeplummerjr** [comment]
This is my exact issue on Ubuntu:
```
>  require('node-opencl/examples/saxpy');
SAXPY with vector size: 524288 elements
using device: GeForce GTX 1080 Ti
node: relocation error: /home/usergpu/node_modules/node-opencl/build/Release/opencl.node: symbol clCreateCommandQueueWithProperties, version OPENCL_2.0 not defined in file libOpenCL.so.1 with link time reference
```

**Comment by robertleeplummerjr** [comment]
I can provide ssh locally, if you like.

**Comment by mikeseven** [comment]
Yes there are a few changes in CL 2.0. I'll write the shim so it's available in js and native.

More importantly I have to update the code to latest nodejs native API as some callbacks don't work as expected in recent versions. Same issue as with node-webgl.

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
Do you know what turnaround time will be? No rush.

**Comment by robertleeplummerjr** [comment]
Also: this is so exciting!

**Comment by mikeseven** [comment]
Hopefully over Thanksgiving ;-)

--mike

________________________________

**Comment by TheLarkInn** [comment]
Bumping for status. I'm very very very interested in this endeavor especially for NodeJS core, and our library [webpack](https://github.com/webpack/webpack) for parallel processing.

**Comment by robertleeplummerjr** [comment]
@mikeseven did you get around to it yet?

**Comment by mikeseven** [comment]
I haven't tested extensively but it has been in feature branch for a while.

**Comment by robertleeplummerjr** [comment]
Nice: https://github.com/mikeseven/node-opencl/tree/feature/upgrade_to_NAN_2.8.0
So we just need to test, give feedback here, and based off that, improve and release or just release?

**Comment by mikeseven** [comment]
That would be nice, thanks.
I tested on some machines here but it would be nice to test if there are any issues with your OS.

-- mike

________________________________

**Comment by TheLarkInn** [comment]
Also this is super relevant I think. (I've tried to drive this with some
success) https://github.com/nodejs/node/issues/18423#issuecomment-366217491

**Comment by mikeseven** [comment]
Yes it is interesting but these shaders, like vulkan's are a subset of OpenCL. The benefits is for graphics apps to avoid a context switch and a syntax easier for shaders developers.

-- mike

________________________________

**Comment by robertleeplummerjr** [comment]
the benefit of gpu.js, with node-opencl (planned, in prototype stage) is that you won't have to learn any of it, just plain old good fashion javascript.

😎

**Comment by mikeseven** [comment]
Gpujs uses webgl so yes you can do gpgpu applications but it's mainly geared at graphics applications. Don't expect doing general OpenCL apps with gl shaders, even compute shaders, because it's quite limited.

-- mike

________________________________

**Comment by robertleeplummerjr** [comment]
As a developer of I can say: GPU.js uses webgl as one of its rendering targets, along with javascript, but it will use opencl soon as another: https://github.com/gpujs/gpu.js/issues/52

It'd be easy to add any number of targets that don't support graphics backends.

**Comment by mikeseven** [comment]
To use opencl as backend, it would need to be outside the browser as none support webcl. So you end up with a solution like node-webcl/webgl.

The problem with js is that it doesn't have the types needed. It could be possible to do some sort of typescript though if you are looking for performance, you would prefer using portable syntax and stay in opencl.

That's one reason I stayed out of a DSL in node-webcl or node-opencl. I preferred providing the low level access and let people invent their preferred DSL on top. Gpujs could be one and so are many others for specific verticals. For example, halide for imaging, ispc for rendering...

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
>To use opencl as backend, it would need to be outside the browser

Already done.  GPU.js originally only used [webgl](https://github.com/gpujs/gpu.js/tree/develop/src/backend/web-gl), but now compiles a kernel just fine on [node](https://github.com/gpujs/gpu.js/tree/develop/src/backend/cpu).  So GPU.js has both node support and browser and handles typed arrays like `Float32Array`, just fine.  Because we built it to work in a specific kernel environment, [type detection is actually super easy](https://github.com/gpujs/gpu.js/blob/develop/src/backend/web-gl/function-node.js#L221).

Note: yes [GPU.js actually compiles a javascript kernel, in javascript](https://github.com/gpujs/gpu.js/blob/develop/src/backend/cpu/function-node.js).  This is done for performance, and... well, just because we can.

The javascript is compiled _directly to_ whatever environment we need.  It doesn't currently support an opencl kernel, which is why I started getting involved in this thread in the first place, all we need is a means of computing the numbers, that is easily installable.  After that, the GPU.js library takes care of the rest.  An example would be compiling a kernel similar to how it is done here in your unit tests: https://github.com/mikeseven/node-opencl/blob/master/test/test.kernel.js#L65

In short the javascript is compiled one to one with the subset of c++ environment (webgl or planned opencl, in the future: arrayfire, cuda, or perhaps webcl if it ever materializes).

**Comment by mikeseven** [comment]
Awesome.

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
Tried to install, this is what I get:
```
M86510-4:gpu.js robert.plummer$ npm i git://github.com/mikeseven/node-opencl.git#7d2de1fbddbd914e1baa41e3ed80287c717c583c
npm WARN deprecated pngjs2@2.0.0: pngjs2 has now taken over the original pngjs package on npm. Please change to use pngjs dependency, version 2+.

> node-opencl@0.5.0 install /Users/robert.plummer/20% Time/gpu.js/node_modules/node-opencl
> node-gyp rebuild --msvs_version=auto

gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
  CXX(target) Release/obj.target/opencl/src/addon.o
  CXX(target) Release/obj.target/opencl/src/types.o
  CXX(target) Release/obj.target/opencl/src/common.o
  CXX(target) Release/obj.target/opencl/src/commandqueue.o
  CXX(target) Release/obj.target/opencl/src/context.o
  CXX(target) Release/obj.target/opencl/src/device.o
  CXX(target) Release/obj.target/opencl/src/event.o
  CXX(target) Release/obj.target/opencl/src/kernel.o
  CXX(target) Release/obj.target/opencl/src/memobj.o
  CXX(target) Release/obj.target/opencl/src/pipe.o
  CXX(target) Release/obj.target/opencl/src/platform.o
  CXX(target) Release/obj.target/opencl/src/program.o
  CXX(target) Release/obj.target/opencl/src/sampler.o
  CXX(target) Release/obj.target/opencl/src/svm.o
  SOLINK_MODULE(target) Release/opencl.node
gpu.js@1.0.0-rc.8 /Users/robert.plummer/20% Time/gpu.js
+-- bluebird@3.5.0  extraneous
+-- node-opencl@0.5.0  invalid (git://github.com/mikeseven/node-opencl.git#7d2de1fbddbd914e1baa41e3ed80287c717c583c)
| +-- chai@4.1.2
| | +-- assertion-error@1.1.0 
| | `-- type-detect@4.0.8 
| +-- get-pixels@3.3.0
| | `-- omggif@1.0.9 
| `-- nan@2.8.0 
`-- node-pre-gyp@0.6.36 

M86510-4:gpu.js robert.plummer$ 
```

**Comment by robertleeplummerjr** [comment]
I ran it once more, and this time no warnings:
```
M86510-4:gpu.js robert.plummer$ npm i --save git://github.com/mikeseven/node-opencl.git#7d2de1fbddbd914e1baa41e3ed80287c717c583c
npm WARN deprecated pngjs2@2.0.0: pngjs2 has now taken over the original pngjs package on npm. Please change to use pngjs dependency, version 2+.

> node-opencl@0.5.0 install /Users/robert.plummer/20% Time/gpu.js/node_modules/node-opencl
> node-gyp rebuild --msvs_version=auto

gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
  CXX(target) Release/obj.target/opencl/src/addon.o
  CXX(target) Release/obj.target/opencl/src/types.o
  CXX(target) Release/obj.target/opencl/src/common.o
  CXX(target) Release/obj.target/opencl/src/commandqueue.o
  CXX(target) Release/obj.target/opencl/src/context.o
  CXX(target) Release/obj.target/opencl/src/device.o
  CXX(target) Release/obj.target/opencl/src/event.o
  CXX(target) Release/obj.target/opencl/src/kernel.o
  CXX(target) Release/obj.target/opencl/src/memobj.o
  CXX(target) Release/obj.target/opencl/src/pipe.o
  CXX(target) Release/obj.target/opencl/src/platform.o
  CXX(target) Release/obj.target/opencl/src/program.o
  CXX(target) Release/obj.target/opencl/src/sampler.o
  CXX(target) Release/obj.target/opencl/src/svm.o
  SOLINK_MODULE(target) Release/opencl.node
gpu.js@1.0.0-rc.8 /Users/robert.plummer/20% Time/gpu.js
+-- bluebird@3.5.0  extraneous
+-- node-opencl@0.5.0  (git://github.com/mikeseven/node-opencl.git#7d2de1fbddbd914e1baa41e3ed80287c717c583c)
`-- node-pre-gyp@0.6.36 
```

**Comment by robertleeplummerjr** [comment]
I believe that means, we are release ready?

**Comment by mikeseven** [comment]
Build should be fine. How about run? Which OSare you using?

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
When I try one of the examples, I get `cl.createContextFromType is not a function`, I'm assuming I'm not doing something right.  I did try just building a kernel, without running it, and it ran perfectly.

**Comment by mikeseven** [comment]
This a cl 2.0 function. Could it be that your machine is opencl 1.x?

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
It is, sadly: https://support.apple.com/en-us/HT202823

Is there v1 example you could point me to?

**Comment by mikeseven** [comment]
Yes mac is stuck on 1.2 even more sad that 2.0 driver does exist for the GPU  they use.

Fear not, CreateWithType is merely syntactic sugar from 1.2. Ok gotta update the samples to use the right method for 1.2. Same for few others. It will be something like if (opencl.version_2)...

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
fantastic!

**Comment by robertleeplummerjr** [comment]
any updates?  Things are moving fast, and want to be able to use soon.

**Comment by mikeseven** [comment]
This weekend for sure...after skiing ;-)

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
nice

**Comment by mikeseven** [comment]
done. Tested on Mac.

On Feb 24, 2018, at 7:04 PM, Robert Plummer <notifications@github.com<mailto:notifications@github.com>> wrote:


nice

**Comment by robertleeplummerjr** [comment]
It builds, but when I run unit tests (which are fantastically written) I get 72 failures, and then then `error Command failed with exit code 72.`.  But a lot of them pass.  Is that normal?  If it is normal, can we get labels perhaps on the tests that identify them as to fail under what conditions?

**Comment by robertleeplummerjr** [comment]
fyi: all examples work 100%
I'd say at the very release, you have a release candidate.

**Comment by mikeseven** [comment]
Yes I corrected the examples a couple of weeks ago and started working on tests. The problem is that some fails on some OS (mainly Mac) but not on others or on some integrated GPU  (e.g. Intel) vs discrete (Nvidia, AMD).
I have added some added some better detection in tests to support these cases and made some modifications but it's not complete.

It has always been a difficult issue as drivers improve. Some features that used to not work suddenly work. Some work via hacks, which is something you definitely don't want to use in your code.

The tests aren't meant to test the node-opencl API which is rather simple. Instead they test opencl driver as a helper to figure out what it really supports and if really compliant to spec.

Today, most vendors are rather good with the exception of Apple, even on latest Macs.

As a guiding rule, stay compliant to opencl 1.2 on Apple. For others, stay compliant to 2.0 as above is still murky.

--mike

________________________________

**Comment by robertleeplummerjr** [comment]
What do you think timeline will be on the next release that allows for opencl 1.2 on Apple?

**Comment by mikeseven** [comment]
It's already working on apple. I test it on high sierra. It's just some features they don't seem to support. In other words the issue is not node-opencl but Apple driver.

--mike

________________________________
